### ISSUE TITLE
Brick and variable selectors

### ISSUE THREAD

**Issue by bartvm** [bartvm/blocks#146]
This is building on the discussions in #140 and #141. Sorry for the long text, but since this is a pretty important part of the framework, let's not leave room for misunderstanding!

I agree with https://github.com/bartvm/blocks/pull/141#discussion_r23456799: Let's use roles such as `BIASES` and `WEIGHTS` to select the parameters we want, and then calculate the weight norms and add that to the cost separately as a regularizer.

However, I think I've put my finger on what I find confusing in the current setup: We are mixing brick selection with variable selection. The selectors from @janchorowski are great, but they are limited by the fact that they operate on the brick level, which means they don't support multiple applications of a single brick, which matters if we want to select input, output or intermediary variables (i.e. variables that do not only depend on the parameters). Take this (artificial) example:
```python
in = tensor.vector()
linear = Linear(input_dim=10, output_dim=10)
tanh = Tanh()
for i in range(3):
    activations = linear.apply(in)
    out = tanh.apply(activations)
    in = out
Linear(input_dim=10, output_dim=20).apply(out)
linear = Linear(input_dim=20, output_dim=20)
for i in range(3):
    activations = linear.apply(in)
    out = tanh.apply(activations)
    in = out
```
I would like to be able to do something along the lines of
```python
cg = ComputationGraph([out])
activations = cg.get_variables(
    predicate=lambda var: VariableRole.OUTPUT in var.tag.roles and
    isinstance(var.tag.brick, Linear))
```
The problem with parameters is that there are two ways of selecting them: Through bricks (because there is only one set of parameters per brick) or through the computation graph. For the sake of simplicity I would argue that our policy (but its' that, just a policy) should be to always *use the computation graph to select variables, and to use brick selectors to select bricks*.

Given what I said above, I wonder if we still need `Annotation` class suggested in https://github.com/bartvm/blocks/issues/140#issuecomment-71096989. Currently, I imagine the following:

* On the brick level:
  * Parameters should be tagged with their roles for easy selection i.e. `VariableRole.WEIGHTS` and `VariableRole.BIASES`
  * We can attach auxiliary variables to the brick object directly. We cannot envision all possible regularizations, so we don't add those, but we can add suggested monitoring channels (which, ironically, would probably include the weight norms). These could be tagged with `VariableRole.MONITOR` for easy selection as well. (We need to implement this, because right now, if an auxiliary variable only depends on the parameters, we construct a new duplicate one each time the brick is applied.)
* On the application call level:
  * Here we add auxiliary variables (basically, suggested monitoring channels) of intermediate results. They too can be tagged with whatever semantics we want. This is also were `VariableRole.INPUT` and `VariableRole.OUTPUT` are tagged.

We can combine variable and brick selection in cases where we want to filter our variable selection to specific bricks e.g. if we want to get all the selected monitoring channels from a particular subset of bricks:
```python
# I'm not very familiary with the selector syntax, but imagine a complicated selector here!
cg.get_variables(predicate=
    lambda var: VariableRole.MONITOR in var.tag.roles and
    var.tag.brick in Selector(mlp1).select('//')
```
But if we want to change the initialization of a set of bricks, we don't touch the computation graph of course
```python
for brick in Selector(mlp1).select('//'):
    brick.weights_init = Orthogonal()
```

**Comment by rizar** [comment]
May I start with a quick question to your example: it seems to me that there is no way to access output variables with the `Selector`. It simply does not have access to them. The example with ` cg.get_variables` you have above can only be done by using the computation graph, can't it?

**Comment by bartvm** [comment]
Yeah, correct, it's just to show that the brick selector doesn't work for cases such as input/output selection and selection of auxiliary variables of application calls.

**Comment by rizar** [comment]
Okay, let's continue in the form of a dialogue. You write: 

> We can attach auxiliary variables to the brick object directly.

which might be sometimes helpful. But then it will be only possible to select using the Selector. Doesn't it contradict your policy to

> use the computation graph to select variables, and to use brick selectors to select bricks.

?

**Comment by bartvm** [comment]
No, just like the `ComputationGraph._get_variabels` function recognises `ApplicationCall` instances and adds the auxiliary variables attached to those, it could do exactly the same for auxiliary variables attached to `Brick` instances. 
```python
if hasattr(current.tag, 'brick'):
    brick = current.tag.brick
    if brick not in self.bricks:
        self.bricks.add(brick)
        for av in brick.auxiliary_variables:
            recursion(av)
```

**Comment by rizar** [comment]
Okay, I see, you want the computation graph to collect those. In practice that would look a bit more involved than in your code example. because variables currently do not have direct links to the bricks, they only reference application calls and those in turn reference bricks. I am not sure I like it: these variables are not parts of computation graphs, they are simply attributes of bricks. If we have such variables, Jan's new selector seems to be better a way to select them.

Also it does not fully eliminate the need for `Annotation` objects, because in your proposal there is no way to add an auxiliary variable or an update without having a brick that does it. But the `Annotation` thingy  can be postponed, it does not have to be there in the very first release.

**Comment by bartvm** [comment]
But we already collect them regardless whether they are part of the computation graph, because we do exactly the same for the auxiliary variables attached to `ApplicationCall`, and those aren't necessarily part of the actual computation of the outputs either. We could create two sets of variables in the computation graph though: Actual variables that are part of the computation and auxiliary variables attached to bricks and application calls (not necessarily disjoint).

The main thing I don't like is that we can't use Jan's selector to select auxiliary variables from application calls, so we end up with a dual system: Use brick selectors to select auxiliary variables attached to bricks (e.g. weight norms) but use variable selectors to select auxiliary variables attached to application calls (e.g. mean activation). Why not do all of them in the same place? Right now it's possible anyway if you try hard enough by filtering the variables attached to the computation graph by the value of `var.tag.application_call.brick`, we would just make it a bit more accessible.

**Comment by bartvm** [comment]
P.S. I see your point about the `Annotation`, and think it's a good idea.

**Comment by rizar** [comment]
Whereas the variables collected from the `AppllicationCall` objects are not necessarily part of the computation graph of the cost as you have rightly mentioned, one can say that they do belong to the computation graph because they were created in the same context as the those variables that belong to the CG literally, and because they are data-dependent. The auxiliary variables you propose to add have very weak relation to the computation graph and in my vision of the world belong to bricks only.

Currently I feel that:
- everything created during calling application methods belongs to the CG and is handled by CG
- parameters are a kind of exception because they belong to both
- the additional brick auxiliaries you propose belong to bricks only

I understand your concern about the dual system, but I think we should not sacrifice conceptual clarity to make the scripts one line shorter. The quantities, such the norm of weights or sparsity, serve for diagnostic of the bricks "health", but they are agnostic to the exact way the brick is optimized, and thus, in my view, should not be part of the CG.

**Comment by rizar** [comment]
And one more point: I would like the `ComputationGraph` to be bricks-agnostic, which is not the case if it has to recurse into bricks.

Then the framework would be structured as follows:

1. Bottom level: computation graph, annotations, variable roles, utility routines and popular regularizations (dropout, weight noise, L2).

2. Component level: bricks + selector, algorithm, dataset, log. All five components use the level 1 but do not user each other.

3. User level: main loop and extensions. Can use all the levels below.

Could be a nice picture!

**Comment by bartvm** [comment]
For me it's not about sacrificing conceptual clarity though, but about gaining it. I find the distinction between e.g. weight norms belonging to bricks, whereas e.g. the mean activation belongs to the CG, very arbitrary.

As a user I have one of two goals in mind: I want to select variables that match A, B or C to add to my monitors/regularize them/use them for something else. Or, I want to configure bricks so and so to initialize differently. As such, I don't care much whether the variables I am looking for are data-dependent or not, and to which part of the library they belong. I also don't want to be collecting half of my monitor channels from the computation graph, while collecting the other half from the bricks. This gets very confusing, especially with cases such as parameters, which kind of belong to both.

As a user, and for me personally, the distinction is between what I am trying to retrieve: Variables can be retrieved one way, bricks the other. There is absolute clarity for everyone about which one which. I understand your distinction, but I don't see what we gain by enforcing it. For us, we can just redefine `ComputationGraph` to encapsulate a Theano graph *including* annotations provided by the Blocks framework if applicable. That means it holds as auxiliary variables anything that was attached to `Annotation`, `ApplicationCall` and `Brick` instances.

**Comment by bartvm** [comment]
Cross-posting again :)

I agree that the `ComputationGraph` should technically be agnostic, but I consider it similar to the `ApplicationCall`. It doesn't rely on any particular interface of bricks. If simply finds them, it stores the auxiliary variables and updates it finds in a separate list of variables, if it doesn't find them (because there aren't any) it functions just as well.

Alternatively, we can make the brick attach a `Annotation` object to its input/output variables. The main reason I would attach the brick is simply because it's another attribute the user can filter variables by, and a very useful one. However, at no point does it rely on the interface whatsoever; as far as `ComputationGraph` is concerned, `brick` is just an attribute that holds auxiliary variables, just like `application_call`.

**Comment by rizar** [comment]
Okay Bart, let's imagine you are a machine learning novice that has not been used to PyLearn2. You start using blocks and here comes the need to monitor the properties of your model. Would you start to look for them in the computation graph when those are simply properties of your components, your bricks? I think that you would not. Because of PyLearn2 you are used to grab all the expressions you care from a so-called model, which is a hybrid of what we call CG and our bricks hierarchy. But PyLearn2 is not our gold standard.

In a very abstract pseudo-code the monitoring related part of blocks-based script can look like that:
```
expressions = collect model properties from bricks
expressions += collect diagnostics of the information flow through your model form the CG
expressions += diagnostics of your optimization algorithm
```

I think that you mainly want to merge the first two because this the way it was done in PyLearn2, but for a newbie those would be very different things.

**Comment by bartvm** [comment]
I actually think it has little to do with Pylearn2, but more about the way I see Blocks: A Theano toolbox that helps us build graphs, not one that abstracts "models" into "bricks". As such, after I built my annotated graph with bricks, I want to say: "Forget about those bricks I used, they built me a nicely annotated graph, what useful things can this graph tell me? Tell me what I can monitor! Tell me all possible things I can use as costs, tell me what variables are biases, and which ones are weights, and which ones are inputs to a linear transformation, etc."

So my workflow would look like
1. Use bricks to build an annotated computational graph, use selectors to configure bricks intelligently
2. Use these annotations through the `ComputationGraph` interface to collect all the Theano variables I care about, whatever role they play. If I want to, I can use bricks to filter the variables I want, but I don't need to; by this time bricks are optional. We are simply dealing with an incredibly powerful Theano graph, which for each variable can tell me whether it's an input, output or parameter, and whether it is a weight, bias or something else. It can tell me what other auxiliary variables I might be interested in to monitor, etc.
3. Use Blocks' main loop, monitoring, etc. to train the Theano variables I go from the computation graph; again, no bricks required.

This is much easier to explain to a newbie: Look, you know Theano? Well, try building a CG using bricks this time, and we add all sorts of nifty annotations to your variables that allow you to select whatever you want. No more need to pass around and keep track of the dozens of Theano variables you want to use. Simply ask the graph for all weights, or anything that can be monitored, or something crazy like the 3rd output and the weight norm of that linear transformation you applied 4 times, and this graph will give it to you!

That sounds easier than explaining to a newbie: Now that you have built your model, you will need to work with two parallel data structures. Based on the role the variable you are looking for plays, you will need to use one of these two parallel data structures to track it down. As a newbie I am lost now: The "information flow" clearly causes my weights to change, but yet their norm is considered fundamentally different from the activations caused by that same data through the same brick? And I can get all the parameters of my model by doing `var for var in cg.get_shared_variables() if VariableRole.PARAMETER in var.tag.role`, so why does the documentation suddenly suggest that I use that brick selection language?

I'm not saying that your distinction is arbitrary, but I do think that for a newbie it is a more abstract distinction than separating bricks and Theano variables. It also helps us achiever our goal of keeping the bricks layer minimal and out of sight when not needed, and not turn bricks into a Pylearn2 model substitute.

**Comment by rizar** [comment]
Okay Bart, your point is also nice and clear. Let's look for a compromise: the think I dislike most in your original proposal is that the CG accesses bricks and has handles to them. So let's just have all the auxiliary variables in the `ApplicationCall`'s. The auxiliaries variables from the bricks can be automatically added to the `ApplicationCall` by the `Application` object. I don't see a big problem if they are added to the CG several times, we can easily filter that and it is completely transparent to the newbie how they made their way to CG. On the other hand, I still will be able to access them with the selector like I want. Does it work for you?

**Comment by bartvm** [comment]
Works for me. I'm working in #141 and will make your suggested changes, I'll ping you when I have something that I think works nicely. Always easier to debate over code when you have an actual implementation!

**Comment by rizar** [comment]
That's true what you say about implementation. I will take a look tomorrow on your intermediate results.

I've just thought that if we stick to having `ComputationGraph` brick-agnostic, that implies that the `cg.get_variables` method should not have brick as a keyword argument. At first I thought this might be a problem, but now it seems I see a way through. The `get_variables` method should limit its interface to accepting a name, roles and a predicate, and we can have a nice family of parametrized predicates to use. Those predicate might also be useful outside of the `ComputationGraph`.

A quick illustration of what I mean:
````python
# Return all variables with the name 'W': I don't know who needs that, but anyways
cg.get_variables(name='W')
# Return all weights.
cg.get_variables(roles=

**Comment by rizar** [comment]
Wait! It is not finished yet! Don't read it!

**Comment by rizar** [comment]
OK, now I am done! Just pressed the button accidentally before finished the post.

That's true what you say about implementation. I will take a look tomorrow on your intermediate results.

I've just thought that if we stick to having `ComputationGraph` brick-agnostic, that implies that the `cg.get_variables` method should not have brick as a keyword argument. At first I thought this might be a problem, but now it seems I see a way through. The `get_variables` method should limit its interface to accepting a name, roles and a predicate, and we can have a nice family of parametrized predicates to use. Those predicate might also be useful outside of the `ComputationGraph`.

A quick illustration of what I mean:
````python
# Return all variables with the Theano name 'W': I don't know who 
# needs that, but anyways
cg.get_variables(theano_name='W')
# Return all parameters.
cg.get_variables(roles=[PARAMETER])
# Return all weights that belong to a brick
cg.get_variables(roles=[WEIGHTS], predicate=VariableFilter(brick=the_brick))
# Return the output "attention_weights" of the "do_apply" method of an attention brick
cg.get_variables(name="attention_weights", roles=[OUTPUT], 
    predicate=VariableFilter(brick=attention, application=attention.do_apply))
```

Looking back at what I have written it seems like I want to have the variable selection logic (not to be confused with brick selection logic) to be separated from the computation graph, which purpose is simple: flatten the computation graph and extract whatever useful is in it.  The selection logic will necessarily have to depend on everything.

**Comment by bartvm** [comment]
Yes, that's pretty much what I had in mind! I realise now that I didn't mention it again since the opening post, but yes, I think `get_variables` should just take a predicate as input (see the second code block in my opening post). That predicate can then optionally depend on variables having application calls and bricks attached, but doesn't need to.

I hadn't thought as far yet as abstracting predicates/filters into their own classes, but I guess that is a natural extension which would save us from writing lengthy `if hasattr(var.tag, 'application_call') and ...` statements all the time.

We'll see how far I get before tomorrow, but I think this could be nice :)

**Comment by bartvm** [comment]
Closed through #141.
